### ISSUE TITLE
Add possibility to stream data

### ISSUE THREAD

**Issue by sirex** [foutaise/texttable#59]
It would be nice to have a possibility to stream a large table without consuming too much memory.

For example `table.draw` could accept a generator and in that case could return another generator, that would yield table rows one by one. If you think, that this is a good idea, I might add this functionality.

Here is an example code, how I imagine it:

```python
rows = ([i, i+1] for i in range(10**9))
table.Texttable()
table.add_rows([['i', 'i+1']])
for chunk in table.draw(rows):
    print(chunk, end='')
```

**Comment by foutaise** [comment]
Hi @sirex,

Thanks for your feedback.

To be honest, I don't see the need for this.

Being concerned about memory consumption while generating an ASCII table
suppose that you have a fairly big amount of data to render. But ASCII tables
are not suited to render large amount of data. It's quickly unreadable,
too wide/tall, etc...

That sounds like premature optimization to me. Yes, me might
implement support for generators, but is there so many use case that
would benefit from this change? That would imply more code (i.e. potential
bugs), and a more complex API for usage that does not match the purpose
of this library («creating simple ASCII tables»).

Furthermore, rows are scanned multiple times to generate the table. So
using a generator that can be read only once seems useless if we have
to keep data in memory behind the scene. We might provide a generator for
the output, but as mentioned before, I don't see any use case that would justify
the effort.

From my point of view, less is more. Generators can be useful in some
circumstances, but not here. Therefore, I close this issue.

Best regards,

**Comment by sirex** [comment]
Use case is pretty simple, I'm working with open data project and implementing support to export data in multiple formats. One of those formats is simple ascii tables. I can't control how much data will be exported, it could be a few rows, but also it can be gigabytes of data.

Currently [as a workaround I paginate table](https://github.com/atviriduomenys/spinta/blob/master/spinta/commands/formats/asciitable.py#L38-L55), where each page contains 100 rows. I use pagination to make sure, that a large dataset would not take up all RAM resulting in a unresponsive server.

Why anyone would wan't to export data in ascii format?

- Personally I use it to quickly investigate data, I just dump all data to a file, and then open it with `less` and this way I can quickly look around even if data takes gigabytes of disk space.
- Another use case, while exporting data through a web api, I use an asyncio web framework, which can build response in small chunks of data. Currently that is not possible, because texttable would block all requests until it finished rendering **all** data.
- Another use case, if I wan't to look at a large dataset, by piping it through `less`, I want to see results immediately while I consume stdout, currently I would have to wait a long time, while whole table is rendered or I have to add some workarounds like pagination.
- If I made a mistake and forgot to add limit or correct filters, I want to see that in the output immediately and interrupt if I see, that I get too much data, instead of waiting, without seeing any output.

> That sounds like premature optimization to me. Yes, me might implement support for generators, but is there so many use case that would benefit from this change?

I already provided multiple use cases. And I think it's not a premature optimization, but a bug. Let's imagine if PostgreSQL would use this library for displaying query results it would simply be not usable, because large results would simply make whole REPL unresponsive.

> more complex API for usage that does not match the purpose of this library («creating simple ASCII tables»).

As you can see from my example, API did not changed at all, then only addition is that `draw` can be used as a generator, if a generator argument is provided. I that, that is a very small addition and the change is backwards compatible.

Purpose stays the same, unless you want to explicitly add «creating simple ASCII tables _limited to small amounts of data_».

> Furthermore, rows are scanned multiple times to generate the table. So using a generator that can be read only once seems useless if we have to keep data in memory behind the scene.

Yes, it is enough to only do that on a small sample to get all the parameters for rendering, and then all other rows can be rendered iteratively. You can take a [slice](https://docs.python.org/3.7/library/itertools.html#itertools.islice) of data, convert it to list and do multiple scans there, then you can continue consume generator.

Anyway, let me know if you really want keep this library restricted only for a small datasets. If that is what you want, then I will have to look for another library.
